storagemarket/impl/providerstates/provider_fsm.go

```diff
@@ -39,7 +40,23 @@ var ProviderEvents = fsm.Events{
 			return nil
 		}),
 	fsm.Event(storagemarket.ProviderEventDataTransferInitiated).
-		From(storagemarket.StorageDealWaitingForData).To(storagemarket.StorageDealTransferring),
+		From(storagemarket.StorageDealWaitingForData).To(storagemarket.StorageDealTransferring).
+		Action(func(deal *storagemarket.MinerDeal, channelId datatransfer.ChannelID) error {
+			deal.TransferChannelId = &channelId
+			return nil
+		}),
+
+	fsm.Event(storagemarket.ProviderEventDataTransferRestartFailed).
+		From(storagemarket.StorageDealProviderTransferRestart).To(storagemarket.StorageDealFailing).
+		Action(func(deal *storagemarket.MinerDeal, err error) error {
+			deal.Message = xerrors.Errorf("error restarting data transfer: %w", err).Error()
+			return nil
+		}),
+
+	fsm.Event(storagemarket.ProviderEventDataTransferRestarted).
+		FromMany(storagemarket.StorageDealWaitingForData, storagemarket.StorageDealProviderTransferRestart).To(storagemarket.StorageDealTransferring).
```

Comment:
What's the scenario where this is called in StorageDealWaitingForData?

Reply:
@hannahhoward 

If the provider crashes after the channel is open on it's side but before seeing the corresponding event in the `ProviderDataTransferSubscriber`,  it will never enter the `StorageDealTransferring` state.